### PR TITLE
Restructured Toggle to use style variables.

### PR DESCRIPTION
**Pull request by khmakoto**
#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [x] Include a change request file using `$ npm run change`

#### Description of changes

Restructured Toggle to use styled variables.

#### Focus areas to test

(optional)

### CODE REVIEW

File: packages/experiments/src/components/Toggle/Toggle.styles.ts [hunk 1]
```diff
@@ -1,222 +1,229 @@
...
 };
+
+type IProcessedVariables<T> = { [P in keyof T]-?: IProcessedVariables<T[P]> };
+
+function _processVariables<T>(partialVariables: T, customVariables?: T): IProcessedVariables<T> {
+  // tslint:disable-next-line:no-any
+  const result = customVariables ? merge({}, partialVariables, customVariables) : partialVariables;
+
+  return result as IProcessedVariables<T>;
+}
```
Comment (khmakoto) [inline packages/experiments/src/components/Toggle/Toggle.styles.ts]: @JasonGore @dzearing 

I copied this from `Button.styles.ts`.

I know it should probably go into `Utilities` as something all components will be using, but wanted your input on where exactly before moving it.

File: packages/experiments/src/components/Toggle/Toggle.styles.ts [hunk 2]
```diff
@@ -1,222 +1,229 @@
...
+      disabled && checked && getToggleStylesFromState(toggleVariables.checkedDisabled)
+    );
+  }
+
+  return concatStyleSets({ root: getFocusStyle(theme) }, getToggleStylesFromVariant(), { root: className });
 };
+
+type IProcessedVariables<T> = { [P in keyof T]-?: IProcessedVariables<T[P]> };
+
+function _processVariables<T>(partialVariables: T, customVariables?: T): IProcessedVariables<T> {
```
Comment (khmakoto) [inline packages/experiments/src/components/Toggle/Toggle.styles.ts]: @JasonGore @dzearing

I copied this from Button.styles.ts.

I know it should probably go into Utilities as something all components will be using, but wanted your input on where exactly before moving it.

File: packages/experiments/src/components/Toggle/docs/ToogleDonts.md [hunk 3]
```diff
@@ -1 +0,0 @@
-- Don’t use a Toggle if the user will have to do something else or go somewhere else in order to experience its effect. If any extra step is required for changes to be effective, you should use a checkbox and corresponding "Apply" button instead of a Toggle.
```
Comment (khmakoto) [inline packages/experiments/src/components/Toggle/docs/ToogleDonts.md]: Realized these were not needed as we could point to the ones already in the OUFR package.

File: packages/experiments/src/components/Toggle/Toggle.state.ts [hunk 4]
```diff
@@ -18,10 +20,17 @@ export class ToggleState extends BaseState<IToggleProps, IToggleViewProps, ITogg
       checked: !!props.defaultChecked,
       text: !!props.defaultChecked ? props.onText : props.offText,
       onChange: this._noop,
-      onClick: this._onClick
+      onClick: this._onClick,
+      toggleButton: this._toggleButton
```
Comment (kenotron) [inline packages/experiments/src/components/Toggle/Toggle.state.ts]: If you're gonna have a ref as a prop, you should at least call it a `toggleButtonRef`

File: common/changes/office-ui-fabric-react/master_2018-10-22-23-24.json [hunk 5]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "office-ui-fabric-react",
+      "comment": "Removing some of the deprecated components flaged when turning the tslint deprecation flag to true in the office-ui-fabric-react package.",
```
Comment (JasonGore) [inline common/changes/office-ui-fabric-react/master_2018-10-22-23-24.json]: Seems like this file shouldn't be part of this PR...

File: packages/experiments/src/components/Toggle/Toggle.styles.ts (hunk 2 above)
Comment (JasonGore) [reply packages/experiments/src/components/Toggle/Toggle.styles.ts]: For this PR at least it could go into `experiments/src/utilities/` to remove the duplication with Button and then finalized when one of the controls using it is promoted.

File: packages/experiments/src/components/Toggle/Toggle.types.ts [hunk 6]
```diff
@@ -88,39 +99,46 @@ export type IToggleViewProps = Pick<IToggleProps, 'as' | 'label' | 'ariaLabel' |
...
+    | 'inherit'
+    | 'unset'
+    | 'flex-start'
+    | 'flex-end'
+    | 'center'
+    | 'space-between'
+    | 'space-around'
+    | 'space-evenly'
+    | 'stretch'
+    | undefined;
```
Comment (JasonGore) [inline packages/experiments/src/components/Toggle/Toggle.types.ts]: These values seem like they should be defined somewhere already, or centralized if not...

File: packages/experiments/src/components/Toggle/Toggle.types.ts [hunk 7]
```diff
@@ -88,39 +99,46 @@ export type IToggleViewProps = Pick<IToggleProps, 'as' | 'label' | 'ariaLabel' |
...
+  pillHighContrastBackground?: string;
+  pillHighContrastHoveredBackground?: string;
+  pillHighContrastBorderColor?: string;
+  pillHighContrastHoveredBorderColor?: string;
+
+  thumbBackground?: string;
+  thumbHighContrastBackground?: string;
+  thumbHighContrastBorderColor?: string;
```
Comment (JasonGore) [inline packages/experiments/src/components/Toggle/Toggle.types.ts]: `pill` and `thumb` seem like they could be slots for Toggle. Do you potentially see these style variables moving out of Toggle's interface and become slot style variables?

File: packages/experiments/src/components/Toggle/__snapshots__/Toggle.view.test.tsx.snap [hunk 8]
```diff
@@ -12,6 +12,22 @@ exports[`ToggleView renders toggle correctly 1`] = `
...
+      .ms-Fabric--isFocusVisible &:focus:after {
+        border: 1px solid #ffffff;
+        bottom: 1px;
+        content: "";
+        left: 1px;
+        outline: 1px solid #666666;
+        position: absolute;
+        right: 1px;
+        top: 1px;
+        z-index: 1;
```
Comment (JasonGore) [inline packages/experiments/src/components/Toggle/__snapshots__/Toggle.view.test.tsx.snap]: Hmm what is the cause for these changes? (I would expect this file to be mostly the same...)

File: packages/experiments/src/components/Toggle/Toggle.state.ts [hunk 9]
```diff
@@ -18,10 +20,17 @@ export class ToggleState extends BaseState<IToggleProps, IToggleViewProps, ITogg
...
     };
   }
 
+  public focus = () => {
+    if (this._toggleButton.current) {
+      this._toggleButton.current.focus();
+    }
+  };
```
Comment (JasonGore) [inline packages/experiments/src/components/Toggle/Toggle.state.ts]: This seems a sensible addition but not as part of style variables.. did something trigger the need for this?

Comment (khmakoto) [reply packages/experiments/src/components/Toggle/Toggle.state.ts]: In changing the component to use style variables I realized I had missed this as part of my previous PR where I transitioned Toggle to use Foundation.

File: packages/experiments/src/components/Toggle/Toggle.types.ts [hunk 10]
```diff
@@ -88,39 +99,46 @@ export type IToggleViewProps = Pick<IToggleProps, 'as' | 'label' | 'ariaLabel' |
...
-   * Thumb inside of the pill.
-   */
-  thumb: IStyle;
+export type IToggleStyleVariables = { [PState in IToggleStates]?: IToggleStyleVariablesTypes };
 
-  /**
-   * Text next to the pill.
-   */
-  text: IStyle;
-}
```
Comment (JasonGore) [inline packages/experiments/src/components/Toggle/Toggle.types.ts]: We need to work out with this and Button and with @dzearing whether we keep these interfaces for backwards compatibility or roll out as a breaking change with Fabric 7

File: packages/experiments/src/components/Toggle/Toggle.styles.ts [hunk 11]
```diff
@@ -58,16 +58,18 @@ export const ToggleStyles: IToggleComponent['styles'] = props => {
...
 
       thumbBackground: semanticColors.disabledBackground,
 
       textColor: semanticColors.disabledText,
       textHighContrastColor: 'GrayText'
     },
 
-    styleVariables
-  });
+    ...styleVariables
```
Comment (JasonGore) [inline packages/experiments/src/components/Toggle/Toggle.styles.ts]: Did you confirm that this is correct? If so, can you add a TODO (or fix) Button too?

One thing we need to do is add a bunch of cases for Button and Toggle trying to style every portion of the component, but we can do that in another PR.

Comment (khmakoto) [reply packages/experiments/src/components/Toggle/Toggle.styles.ts]: I did not confirm this. Actually, having it as a separate PR might be a good thing, still need to make a couple of changes here before being able to merge, though.

File: packages/experiments/src/components/Toggle/__snapshots__/Toggle.view.test.tsx.snap (hunk 8 above)
Comment (khmakoto) [reply packages/experiments/src/components/Toggle/__snapshots__/Toggle.view.test.tsx.snap]: I realized what I was doing here. It's fixed, the snapshot should be the same as it was before.

File: packages/experiments/src/components/Toggle/Toggle.types.ts [hunk 12]
```diff
@@ -88,39 +100,35 @@ export type IToggleViewProps = Pick<IToggleProps, 'as' | 'label' | 'ariaLabel' |
...
+
+  textColor?: string;
+  textHighContrastColor?: string;
+}
 
-  /**
-   * Thumb inside of the pill.
-   */
-  thumb: IStyle;
+export type IToggleStyleVariables = { [PState in IToggleStates]?: IToggleStyleVariablesTypes };
```
Comment (dzearing) [inline packages/experiments/src/components/Toggle/Toggle.types.ts]: I think this should be:
```ts
IToggleStyleVariableTypes | (props: IToggleViewProps) => IToggleStyleVariableTypes;
```
Which could be abstracted in a `type` definition.

Comment (dzearing) [reply packages/experiments/src/components/Toggle/Toggle.types.ts]: In other words, exposing the state in the type I think is confusing. States are not necessarily mutually exclusive, and therefore, what defines which state has a higher precedence?

Example:

```ts
{
  enabled: {
    background: 'red'
  },
  checked: {
     background: 'green'
  }
}
```

What color is it?

Comment (JasonGore) [reply packages/experiments/src/components/Toggle/Toggle.types.ts]: I didn't catch this earlier, but `checked` here implies `enabled` and states are mutually exclusive. I think it's unintuitive and something we need to work on
